(CDAP-16353) Introduce ArtifactRepositoryReader interface and implement local and remote versions

wyzhang opened the pull request:
Goal:
- Support ArtifactRepository impl to get artifact detail from either local or remote

Specially this change does:
- Create a new interface ArtifactRepositoryReader and move getArtifact method over from ArtifactRepository
- Implement local and remote version of ArtifactRepositoryReader
- Use local version in all places (e.g. Preview and AppFabric).
- Remote version is unused at the moment. Will switch preview to use it later when storage is set to NOSQL

yaojiefeng commented on cdap-app-fabric/src/main/java/io/cdap/cdap/app/guice/AppFabricServiceRuntimeModule.java:
```diff
@@ -294,11 +297,23 @@ protected void configure() {
+            .annotatedWith(Names.named(NOAUTH_ARTIFACT_REPO))
+            .to(DefaultArtifactRepository.class)
+            .in(Scopes.SINGLETON);
+          expose(ArtifactRepository.class).annotatedWith(Names.named(NOAUTH_ARTIFACT_REPO));
```
you are anyway exposing these bindings, then you don't need to remove these in private module, in the private module, you can just have the ArtifactRepositoryReader binding.

yaojiefeng replied:
also is `ArtifactRepositoryReader` only used by ArtifactRepository? If some other classes wants to use it, will it fail to inject because it is private?

wyzhang replied:
I think we would need to have bind(RepoReader) and bind(Repo) in the same module otherwise RepoReader binding is not visible to bind(Repo). 

Yes. RepoReader is intended to be used by Repo only at this moment. 

Yes. If other classes need to use it. They would do something like 
PrivateModule {
  bind(RepoReader).to(...) 
  bind(SomeClassInterface).to(SomeClassImpl) 
}

yaojiefeng replied:
That is my concern since the purpose of binding is to make injection easier, a private module should only be used if a class is specific to some other classes. If some other classes needs an interface, they need to bind it themselves, that's no different than calling the constructor themselves.

wyzhang replied:
The use case here is using private module to allow different bindings for the same class. As in a follow-up CL, RepoReader class will be bound to local in appfabric and remote for preview when we use local level DB. 

Here is a stackoverflow thread related to this topic
https://stackoverflow.com/questions/16987815/guice-binding-several-objects-with-different-dependencies

In general, if we require class A to always be bound to class AImpl, then yes probably no need to use private module, or use private with expose. But if we need to bind class A to class AImplX or class AImplY depending on some other condition, then private module make the binding easy and cleaner to do.

yaojiefeng commented on cdap-app-fabric/src/main/java/io/cdap/cdap/app/guice/InMemoryProgramRunnerModule.java:
```diff
@@ -32,7 +33,11 @@
 import io.cdap.cdap.common.discovery.ResolvingDiscoverable;
 import io.cdap.cdap.internal.app.runtime.artifact.ArtifactFinder;
 import io.cdap.cdap.internal.app.runtime.artifact.ArtifactManagerFactory;
+import io.cdap.cdap.internal.app.runtime.artifact.ArtifactRepository;
```
nit - unused imports?

wyzhang replied:
Removed

yaojiefeng commented on cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/DefaultPreviewManager.java:
```diff
@@ -345,25 +348,35 @@ protected void configure() {
             .annotatedWith(Names.named(AppFabricServiceRuntimeModule.NOAUTH_ARTIFACT_REPO))
             .to(DefaultArtifactRepository.class)
             .in(Scopes.SINGLETON);
+          expose(ArtifactRepository.class)
```
Same here, you are exposing this binding anyway, why make it private

wyzhang replied:
This way we limit the visibility of RepoReader's binding, meaning others can bind RepoReader differently.

yaojiefeng commented on cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/artifact/ArtifactDescriptor.java:
```diff
@@ -19,17 +19,27 @@
+   * gets serialized when this object needs to be transferred through the network.
+   * The recipient should instantiate a {@link Location} based on the received "locationURI"
+   */
+  private final transient Location location;
```
why do we make this change? Is this class used anywhere in the response? If so, this is changing the API behavior

wyzhang replied:
This PR introduces a REST endpoint for getting artifact detail, thus requiring support for serialization and deserialization. But Location is an interface that cannot be sterilized easily, thus ignoring this field and using location URI instead. The requester would need to convert from Location URI to Location using proper location factory.

yaojiefeng replied:
This means when it get deserialized, the location will be null, which makes this class unreliable, is it necessary to return the entire ArtifactDetail back in the REST call, can we just return a subset of them?

yaojiefeng replied:
This means when it get deserialized, the location will be null, which makes this class unreliable, is it necessary to return the entire ArtifactDetail back in the REST call, can we just return a subset of them?

wyzhang replied:
Correct on location will be null. But the Remote repo reader client side implication inits Location so it returns a complete ArtifactDetail to the caller. 

Actually Location object is one of the key info we need to fetch from remote for preview.

yaojiefeng replied:
Better to put a comment about this, otherwise it will be super confusing to future development

wyzhang replied:
Okay. Added some more to the description.

yaojiefeng commented on cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/artifact/RemoteArtifactRepositoryReader.java:
```diff
@@ -0,0 +1,93 @@
+/**
+ * Implementation for fetching artifact metadata from remote {@link ArtifactHttpHandlerInternal}
+ */
+public class RemoteArtifactRepositoryReader implements ArtifactRepositoryReader {
```
I don't see this class is binded in anywhere, where should it get bind?

wyzhang replied:
Remote repo reader isn't being used anywhere yet. In the next PR, will make preview to use remote repo reader for local levelDB.

yaojiefeng reviewed (commented):
Some questions

yaojiefeng reviewed (commented):
Can you add some comments on why the private module is needed for the reader? Rest LGTM

wyzhang commented:
As for adding comments for private module, let me do that in the next PR where we have different bindings depending on storage implementation. Hopefully it would be a bit more clear why it is done this way.

wyzhang committed:
(CDAP-16353) Introduce ArtifactRepositoryReader interface and implement local and remote versions

Goal:
- Support ArtifactRepository impl to get artifact detail from either local store or remote Appfabric

Specially this change includes:
- Create a new interface ArtifactRepositoryReader and move getArtifact method over from ArtifactRepository
- Implement local and remote version of ArtifactRepositoryReader
- Use local version in all places (e.g. Preview and AppFabric).
- Remote version is unused at the moment. Will switch preview to use it later when storage is NOSQL
